README.md
# N ary tree

[n-ary tree](https://en.wikipedia.org/wiki/M-ary_tree)(also known as `k-ary` or `k-way` tree) implementation in `JavaScript(TypeScript)`.

![Pipeline](https://github.com/lbwa/n-ary/workflows/Pipeline/badge.svg) <img alt="npm (tag)" src="https://img.shields.io/npm/v/n-ary-tree/latest?style=flat-square"> <img alt="npm bundle size" src="https://img.shields.io/bundlephobia/minzip/n-ary-tree?style=flat-square">

<!-- TOC -->

- [N ary tree](#n-ary-tree)
  - [Installation](#installation)
  - [Tree node structures](#tree-node-structures)
    - [Default structure fields](#default-structure-fields)
    - [Examples](#examples)
  - [APIs](#apis)
    - [findNodes](#findnodes)
    - [findPathNodes](#findpathnodes)
    - [findPath](#findpath)
    - [findAllPaths](#findallpaths)
    - [Traversal](#traversal)
      - [levelorder](#levelorder)
      - [preorder](#preorder)
      - [postorder](#postorder)
  - [License](#license)

<!-- /TOC -->

## Installation

- Using npm

```bash
$ npm i n-ary-tree
```

- Using yarn

```bash
$ yarn add n-ary-tree
```

## Tree node structures

User-defined tree node fields have supported by all available methods. Just keep the last parameter is a structure like:

```ts
type TreeNodeFields<N> = Partial<{
  value: keyof N
  children: keyof N
}>
```

### Default structure fields

```ts
{
  value: 'value',
  children: 'children'
}
```

By default, we will use `value` field as the value of tree node, `children` field as the children of tree node.

### Examples

```ts
{
  value: 'val', // or other field string in the tree node
  children: 'descendants' // or other field string in the tree node
}
```

`val` field will be regarded as node actual value, `descendants` field in the tree node will be regraded as node children field.

## APIs

### findNodes

Get all matched nodes.

```ts
export function findNodes<N extends Record<string, any>, V = any>(
  root: N,
  targets: V[],
  fields?: TreeNodeFields<N>
): FindNodesResult<N, V>
```

```ts
import { findNodes } from 'n-ary-tree'

interface DefaultTreeNode {
  value: number
  children?: DefaultTreeNode[]
}

const tree: DefaultTreeNode = {
  value: 1,
  children: [
    {
      value: 11,
      children: [
        {
          value: 111
        }
      ]
    },
    {
      value: 12,
      children: [
        {
          value: 121
        }
      ]
    }
  ]
}

findNodes(tree, [111])
// -> {
//      nodes: [{ value: 111 }],
//      values: [111]
//    }
findNodes(tree, [111], { value: 'value', children: 'children' })
// Equivalent to findNodes(tree, [111]), { value: 'value', children: 'children' }
// is default preset.
```

### findPathNodes

Get nodes sequences if tree path exist.

```ts
import { findPathNodes } from 'n-ary-tree'

findPathNodes(tree, [1, 11, 111])
/**
 * [
 *   { value: 1, children: ... },
 *   { value: 11, children: ... },
 *   { value: 111, children: ...}
 * ]
 */

findPathNodes(tree, [1, 9])
/**
 * [
 *   { value: 1, children: ... }
 * ]
 */
```

### findPath

Find the latest matched path.

```ts
import { findPath } from 'n-ary-tree'

findPath(tree, 121)
/**
 * [
 *   { value: 1, children: ... },
 *   { value: 12, children: ... },
 *   { value: 121, children: ...}
 * ]
 */
findPath(tree, 22) // []
```

### findAllPaths

Find all matched paths.

```ts
import { findAllPaths } from 'n-ary-tree'

const tree = {
  value: 1,
  children: [
    {
      value: 111
    },
    {
      value: 11,
      children: [
        {
          value: 111
        }
      ]
    },
    {
      value: 12,
      children: [
        {
          value: 121
        }
      ]
    }
  ]
}

findAllPaths(tree, [111])
/**
 * [
 *    [
 *      { value: 1, children: ... },
 *      { value: 111, children: ... },
 *    ],
 *    [
 *      { value: 1, children: ... },
 *      { value: 11, children: ... },
 *      { value: 111, children: ...}
 *    ]
 * ]
 */
```

### Traversal

We have support 3 kinds of common tree traversal methods: [level-order(BFS)](https://en.wikipedia.org/wiki/Tree_traversal#Breadth-first_search_/_level_order), [pre-order(DFS)](<https://en.wikipedia.org/wiki/Tree_traversal#Pre-order_(NLR)>), [post-order(DFS)](<https://en.wikipedia.org/wiki/Tree_traversal#Post-order_(LRN)>).

#### levelorder

```ts
const tree = {
  val: 1,
  children: [
    {
      val: 2
    }
  ]
}
levelorder(tree, { value: 'val' })
// [[1], [2]]
```

#### preorder

```ts
const tree = {
  value: 1,
  descendants: [
    {
      value: 2
    }
  ]
}
preorder(tree, { children: 'descendants' })
// [1, 2]
```

#### postorder

```ts
const tree = {
  val: 1,
  descendants: [
    {
      val: 2
    }
  ]
}
postorder(tree, { value: 'val', children: 'descendants' })
// [2, 1]
```

## License

[MIT](./LICENSE) © [Liu Bowen](https://github.com/lbwa)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.4.0](https://github.com/lbwa/n-ary/compare/v0.3.0...v0.4.0) (2020-11-14)


### Features

* **find:** find all/last path(s) with a specific value ([62f6e05](https://github.com/lbwa/n-ary/commit/62f6e058b39fa6712f74f97c7bcba39ee4a95539))


### Bug Fixes

* ignore undefined, null type value ([f7cd6d4](https://github.com/lbwa/n-ary/commit/f7cd6d445c283cabe40d6c2da6bf439e35598eef))

## [0.3.0](https://github.com/lbwa/n-ary/compare/v0.2.2...v0.3.0) (2020-10-20)


### Features

* **findPathNodes:** exit children bfs when no more target value ([dd8533d](https://github.com/lbwa/n-ary/commit/dd8533dbf1254e1f4090a9a807b181323eaf01e4))
* **findPathNodes:** impl findPathNodes ([b2a7e70](https://github.com/lbwa/n-ary/commit/b2a7e70492794ec1ee9e02a413729f9208209aa2))

### [0.2.2](https://github.com/lbwa/n-ary/compare/v0.2.1...v0.2.2) (2020-10-19)


### Bug Fixes

* **findNodes:** shouldn't judge answer length in dfs ([4d01077](https://github.com/lbwa/n-ary/commit/4d010777abf463485d5dc0affa69f570bbd41bec))

### [0.2.1](https://github.com/lbwa/n-ary/compare/v0.2.0...v0.2.1) (2020-10-17)


### Bug Fixes

* format npm description ([2c02422](https://github.com/lbwa/n-ary/commit/2c0242267fca5839b9c01b7a9b2ae2009993da77))

## [0.2.0](https://github.com/lbwa/n-ary/compare/v0.1.0...v0.2.0) (2020-10-17)


### Features

* rename traversal methods ([f3e2a11](https://github.com/lbwa/n-ary/commit/f3e2a114e9e57ba9bfc46603a941ca3a31393c35))
* **traversal:** impl preorder, postorder, levelorder traversal ([d8c2584](https://github.com/lbwa/n-ary/commit/d8c2584fde422d01c1beeafb7249b03472d0d1b9))


### Bug Fixes

* **traversal:** types ([88f2d6d](https://github.com/lbwa/n-ary/commit/88f2d6db6f2f4b847617b09e512e1bfdab9d906b))


src/find.spec.ts
import { findNodes, findPathNodes, findAllPaths, findPath } from '.'

interface DefaultTreeNode {
  value: number
  children?: DefaultTreeNode[]
}

describe('find* methods', () => {
  const tree: DefaultTreeNode = {
    value: 1,
    children: [
      {
        value: 11
      },
      {
        value: 12,
        children: [
          {
            value: 121
          },
          {
            value: 122
          }
        ]
      },
      {
        value: 13
      },
      {
        value: 14,
        children: [
          {
            value: 141
          },
          {
            value: 142
          }
        ]
      }
    ]
  }

  it('should find nodes with default fields', () => {
    expect(findNodes(undefined as any, [11])).toEqual({ nodes: [], values: [] })
    expect(findNodes(tree, [11])).toEqual({
      nodes: [{ value: 11 }],
      values: [11]
    })
    expect(findNodes(tree, [142])).toEqual({
      nodes: [{ value: 142 }],
      values: [142]
    })
    expect(findNodes(tree, [123])).toEqual({ nodes: [], values: [] })
    expect(
      findNodes(
        {
          value: 1,
          children: new Array(99).fill(undefined).concat({
            value: 12,
            children: new Array(99)
              .fill(undefined)
              .concat({ value: 19 })
              .concat(new Array(99).fill(null))
          })
        },
        [19]
      )
    ).toEqual({ nodes: [{ value: 19 }], values: [19] })
  })

  it('should find nodes with specific fields', () => {
    const tree = {
      label: 'label1',
      descendants: [
        {
          label: 'label11',
          descendants: [
            {
              label: 'label111'
            }
          ]
        }
      ]
    }
    expect(
      findNodes(tree, ['label11'], { value: 'label', children: 'descendants' })
    ).toEqual({ nodes: [tree.descendants![0]], values: ['label11'] })
    expect(
      findNodes(tree, ['label112'], { value: 'label', children: 'descendants' })
    ).toEqual({ nodes: [], values: [] })
  })

  it('should find nodes in the path', () => {
    expect(findPathNodes(undefined as any, [1, 14, 142])).toHaveLength(0)
    expect(findPathNodes(tree, [2])).toEqual([])
    expect(findPathNodes(tree, [1, 99])).toEqual([tree])
    expect(findPathNodes(tree, [1, 14, 10])).toEqual([
      tree,
      {
        value: 14,
        children: [
          {
            value: 141
          },
          {
            value: 142
          }
        ]
      }
    ])
    expect(findPathNodes(tree, [1, 14, 142])).toEqual([
      tree,
      {
        value: 14,
        children: [
          {
            value: 141
          },
          {
            value: 142
          }
        ]
      },
      {
        value: 142
      }
    ])
  })

  it('should find path nodes with specific fields', () => {
    const tree = {
      label: 'label1',
      descendants: [
        {
          label: 'label11',
          descendants: [
            {
              label: 'label111',
              descendants: [
                {
                  label: 'label1111'
                }
              ]
            }
          ]
        }
      ]
    }
    expect(
      findPathNodes(tree, ['label11'], {
        value: 'label',
        children: 'descendants'
      })
    ).toEqual([])
    expect(
      findPathNodes(undefined as any, [], {
        value: 'label',
        children: 'descendants'
      })
    ).toEqual([])
    expect(
      findPathNodes(tree, [], {
        value: 'label',
        children: 'descendants'
      })
    ).toEqual([])
    expect(
      findPathNodes(tree, ['label1', 'label11'], {
        value: 'label',
        children: 'descendants'
      })
    ).toEqual([tree, tree.descendants![0]])
    expect(
      findPathNodes(tree, ['label1', 'label112'], {
        value: 'label',
        children: 'descendants'
      })
    ).toEqual([tree])
  })

  it('should find last matched path when we have only one target node', () => {
    const tree = {
      label: 'label1',
      descendants: [
        {
          label: 'label11',
          descendants: [
            {
              label: 'label1111' // duplicated
            },
            {
              label: 'label111',
              descendants: [
                {
                  label: 'label1111' // duplicated
                },
                {
                  label: 'label1112'
                }
              ]
            },
            {
              label: 'label112'
            }
          ]
        }
      ]
    }
    expect(
      findPath(tree, 'label2', { value: 'label', children: 'descendants' })
    ).toEqual([])
    expect(
      findPath(tree, 'label111', { value: 'label', children: 'descendants' })
    ).toEqual([tree, tree.descendants[0], tree.descendants[0].descendants[1]])
    expect(
      findPath(tree, 'label112', { value: 'label', children: 'descendants' })
    ).toEqual([tree, tree.descendants[0], tree.descendants[0].descendants[2]])
    expect(
      findPath(tree, 'label1112', { value: 'label', children: 'descendants' })
    ).toEqual([
      tree,
      tree.descendants[0],
      tree.descendants[0].descendants[1],
      tree.descendants[0].descendants![1].descendants![1]
    ])
    expect(
      findPath(tree, 'label1111', { value: 'label', children: 'descendants' })
    ).toEqual([
      tree,
      tree.descendants[0],
      tree.descendants[0].descendants[1],
      tree.descendants[0].descendants![1].descendants![0]
    ])
  })

  it('should find all matched path when we have only one target node', () => {
    const tree = {
      label: 'label1',
      descendants: [
        {
          label: 'label11',
          descendants: [
            {
              label: 'label1111' // duplicated
            },
            {
              label: 'label111',
              descendants: [
                {
                  label: 'label1111' // duplicated
                },
                {
                  label: 'label1112'
                }
              ]
            },
            {
              label: 'label112'
            }
          ]
        }
      ]
    }
    expect(
      findAllPaths(tree, 'label1111', {
        value: 'label',
        children: 'descendants'
      })
    ).toEqual([
      [tree, tree.descendants[0], tree.descendants[0].descendants[0]],
      [
        tree,
        tree.descendants[0],
        tree.descendants[0].descendants[1],
        tree.descendants[0].descendants![1].descendants![0]
      ]
    ])
  })
})


src/index.spec.ts
import { findNodes } from '.'

interface DefaultTreeNode {
  value: number
  children?: DefaultTreeNode[]
}

interface TreeNode {
  label: string
  descendants?: TreeNode[]
}

describe('N-ary tree', () => {
  it('should find nodes with default fields', () => {
    const tree: DefaultTreeNode = {
      value: 1,
      children: [
        {
          value: 11,
          children: [
            {
              value: 111
            }
          ]
        },
        {
          value: 12,
          children: [
            {
              value: 121
            }
          ]
        }
      ]
    }
    expect(findNodes(tree, [111])).toEqual({
      nodes: [{ value: 111 }],
      values: [111]
    })
    expect(findNodes(tree, [123])).toEqual({ nodes: [], values: [] })
  })

  it('should find nodes with specific fields', () => {
    const tree: TreeNode = {
      label: 'label1',
      descendants: [
        {
          label: 'label11',
          descendants: [
            {
              label: 'label111'
            }
          ]
        }
      ]
    }
    expect(
      findNodes(tree, ['label11'], { value: 'label', children: 'descendants' })
    ).toEqual({ nodes: [tree.descendants![0]], values: ['label11'] })
    expect(
      findNodes(tree, ['label112'], { value: 'label', children: 'descendants' })
    ).toEqual({ nodes: [], values: [] })
  })
})


src/traversal.spec.ts
import { preorder, postorder, levelorder } from './traversal'

describe('N ary tree traversal', () => {
  const defaultTree = {
    value: 1,
    children: [
      { value: 3, children: [{ value: 5 }, { value: 6 }] },
      { value: 2 },
      { value: 4 }
    ]
  }
  const customFieldsTree = {
    val: 1,
    descendants: [
      { val: 3, descendants: [{ val: 5 }, { val: 6 }] },
      { val: 2 },
      { val: 4 }
    ]
  }
  it('preorder with default fields', () => {
    expect(preorder()).toEqual([])
    expect(preorder(defaultTree)).toEqual([1, 3, 5, 6, 2, 4])
  })

  it('preorder with customize fields', () => {
    expect(preorder()).toEqual([])
    expect(
      preorder(customFieldsTree, { value: 'val', children: 'descendants' })
    ).toEqual([1, 3, 5, 6, 2, 4])
  })

  it('postorder with default fields', () => {
    expect(postorder()).toEqual([])
    expect(postorder(defaultTree)).toEqual([5, 6, 3, 2, 4, 1])
  })

  it('postorder with customize fields', () => {
    expect(postorder()).toEqual([])
    expect(
      postorder(customFieldsTree, { value: 'val', children: 'descendants' })
    ).toEqual([5, 6, 3, 2, 4, 1])
  })

  it('levelorder with default fields', () => {
    expect(levelorder()).toEqual([])
    expect(levelorder(defaultTree)).toEqual([[1], [3, 2, 4], [5, 6]])
  })

  it('levelorder with customize fields', () => {
    expect(
      levelorder(customFieldsTree, {
        value: 'val',
        children: 'descendants'
      })
    ).toEqual([[1], [3, 2, 4], [5, 6]])
  })
})


src/find.ts
import { TreeNode, TreeNodeFields } from './types'
import { isDef } from 'tslang-utils'

export type FindNodesResult<N extends TreeNode<V>, V> = {
  nodes: N[]
  values: V[]
}

export function findNodes<N extends TreeNode<V>, V = any>(
  root: N,
  targets: V[],
  fields?: TreeNodeFields<N>
): FindNodesResult<N, V>

export function findNodes<N extends TreeNode<V>, V = any>(
  root: N,
  targets: V[],
  { value = 'value', children = 'children' }: TreeNodeFields<N> = {}
): FindNodesResult<N, V> {
  const values: V[] = []
  const nodes: N[] = []

  if (!root) {
    return { nodes, values }
  }

  function dfs(root: N | undefined) {
    if (!root) {
      return
    }

    const index = targets.indexOf(root[value])
    if (index > -1) {
      values[index] = root[value]
      nodes[index] = root
    }

    if (Array.isArray(root[children]) && root[children].length > 0) {
      root[children].forEach(dfs)
    }
  }
  dfs(root)
  return { nodes, values }
}

export function findPathNodes<N extends TreeNode<V>, V = any>(
  root: N,
  targets: V[],
  fields?: TreeNodeFields<N>
): N[]

export function findPathNodes<N extends TreeNode<V>, V = any>(
  root: N,
  targets: V[],
  { value = 'value', children = 'children' }: TreeNodeFields<N> = {}
): N[] {
  const nodes: N[] = []
  if (!root) {
    return nodes
  }

  const queue: N[] = [root]
  let levelIndex = -1
  while (queue.length > 0) {
    const levelSize = queue.length
    levelIndex++
    for (let i = 0; i < levelSize; i++) {
      const current = queue.shift()!

      if (current[value] === targets[levelIndex]) {
        nodes.push(current)
        if (levelIndex < targets.length - 1) {
          queue.push(...((current[children] as any) || []))
        }
        break
      }
    }
  }
  return nodes
}

function createPathFinder<
  N extends TreeNode<V>,
  R extends N[] | N[][],
  V = any
>(callback: (track: N[], answer?: R) => R) {
  return function pathFinder(
    root: N,
    target: V,
    { value = 'value', children = 'children' }: TreeNodeFields<N> = {}
  ): R {
    const track: N[] = []
    let answer: R = ([] as unknown) as R
    function backtracking(root: N | undefined | null) {
      if (!root) {
        return
      }
      track.push(root)
      if (isDef(root[value]) && root[value] === target) {
        answer = callback(track, answer)
      }
      if (Array.isArray(root[children])) {
        root[children].forEach(backtracking)
      }
      track.pop()
    }
    backtracking(root)
    return answer
  }
}

export function findPath<N extends TreeNode<V>, V = any>(
  root: N,
  target: V,
  treeNodeFields: TreeNodeFields<N> = {}
) {
  return createPathFinder<N, N[], V>(track => [...track])(
    root,
    target,
    treeNodeFields
  )
}

export function findAllPaths<N extends TreeNode<V>, V = any>(
  root: N,
  target: V,
  treeNodeFields: TreeNodeFields<N> = {}
) {
  return createPathFinder<N, N[][], V>((track, answer) =>
    answer!.concat([[...track]])
  )(root, target, treeNodeFields)
}


src/traversal.ts
import { TreeNode, TreeNodeFields } from './types'
import { isDef } from 'tslang-utils'

export function levelorder<N extends TreeNode<V>, V = any>(
  root?: N,
  { value = 'value', children = 'children' }: TreeNodeFields<N> = {}
) {
  const ans: V[][] = []
  if (!isDef(root)) {
    return ans
  }

  const queue: N[] = [root]
  while (queue.length > 0) {
    const levelSize = queue.length
    ans.push([])
    for (let i = 0; i < levelSize; i++) {
      const current = queue.shift()!
      ans[ans.length - 1].push(current[value])

      const nextLevel = current[children]
      if (Array.isArray(nextLevel)) {
        queue.push(...(nextLevel as N[]))
      }
    }
  }
  return ans
}

export const bfs = levelorder

export function preorder<N extends TreeNode<V>, V = any>(
  root?: N,
  { value = 'value', children = 'children' }: TreeNodeFields<N> = {}
) {
  const ans: V[] = []
  if (!isDef(root)) {
    return ans
  }

  function dfs(root: N | undefined, ans: V[]) {
    if (!isDef(root)) {
      return
    }
    ans.push(root[value])
    if (Array.isArray(root[children]) && (root[children] as any).length > 0) {
      ;(root[children] as any).forEach((child: N) => dfs(child, ans))
    }
  }
  dfs(root, ans)
  return ans
}

export function postorder<N extends TreeNode<V>, V = any>(
  root?: N,
  { value = 'value', children = 'children' }: TreeNodeFields<N> = {}
) {
  const ans: V[] = []
  if (!isDef(root)) {
    return ans
  }

  function dfs(root: N | undefined, ans: V[]) {
    if (!isDef(root)) {
      return
    }
    if (Array.isArray(root[children]) && (root[children] as any).length > 0) {
      ;(root[children] as any).forEach((child: N) => dfs(child, ans))
    }
    ans.push(root[value])
  }
  dfs(root, ans)
  return ans
}


perf/index.js
const Benchmark = require('benchmark')
const {
  findNodes: findNodesInProd,
  findPathNodes: findPathNodesInProd
} = require('../dist/n-ary-tree.cjs.production.min')
const {
  findNodes: findNodesInDev,
  findPathNodes: findPathNodesInDev
} = require('../dist/n-ary-tree.cjs.development')

const targets = [142]
const path = [1, 14, 142]
Benchmark.Suite()
  .add(`findNodes in prod`, function() {
    findNodesInProd(tree, targets)
  })
  .add('findNodes is dev', function() {
    findNodesInDev(tree, targets)
  })
  .add('findPathNodes in prod', function() {
    findPathNodesInProd(tree, path)
  })
  .add('findPathNodes in dev', function() {
    findPathNodesInDev(tree, path)
  })
  .on('cycle', function(event) {
    console.log(String(event.target))
  })
  .on('complete', function() {
    console.log('Fastest is ' + this.filter('fastest').map('name'))
  })
  .run({ async: true })

const tree = {
  value: 1,
  children: [
    {
      value: 11
    },
    {
      value: 12,
      children: [
        {
          value: 121
        },
        {
          value: 122
        }
      ]
    },
    {
      value: 13
    },
    {
      value: 14,
      children: [
        {
          value: 141
        },
        {
          value: 142
        }
      ]
    }
  ]
}
